### PR TITLE
Charm Platform Origin and Bundles

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following updates the deploy code to allow for platform to reside in
origin. With this change, it's now possible to correctly wire up the bundle
support and model migration changes. This helps with the wire up of the
architecture to platform so we can effectively run model migrations with
architecture support.

### Architecture

Architecture is wired through the deployment code by using the arch
constraint supplied and pushing it into the charm origin and pushing
that to the server.

There is a hack where the arch sent to the charmhub is hardcoded to
"all" because of a major issue in the API that is currently being resolved.

### Bundle support

Bundle support is wired up so that we call the FindDownloadURL on the
server. The bundle URL along with the origin is passed back so we can
extract it from the client. Originally I was going to create a generic repo
adapter, but it causes even more churn. Instead, I create one locally that
correctly loads the bundle and returns it to the deployer.

### Selector

The selector code that was crafted for downloads was repurposed for
finding the correct bundle URL. It was moved to `charmhub/selector` so
that client and server could use it.

There is a gap in the selector code that will be fixed in another PR that
will fix the issue around revisions and selecting the correct one based on
either the charm URL or the origin.

### Model migration

The changes here just exposed the platform to the model migration.

## QA steps

There are no bundles that have charmhub charms in them. This PR is only
to ensure we get the correct bundle.

```sh
$ export JUJU_DEV_FEATURE_FLAGS=charm-hub
$ juju bootstrap lxd test --config charm-hub-url="https://api.staging.snapcraft.io"
$ juju deploy wiki-simple
Located bundle "ch:wiki-simple-4"
Resolving charm via charmstore: cs:trusty/mysql-55
Resolving charm via charmstore: cs:trusty/mediawiki-5
Executing changes:
- upload charm cs:mysql-55 for series trusty
- deploy application mysql on trusty using cs:mysql-55
- set annotations for mysql
- upload charm cs:trusty/mediawiki-5 for series trusty
- deploy application wiki on trusty using cs:trusty/mediawiki-5
- set annotations for wiki
- add relation wiki:db - mysql:db
- add unit mysql/0 to new machine 0
- add unit wiki/0 to new machine 1
Deploy of bundle completed.

```

### Fallback architecture and series.

```sh
$ export JUJU_DEV_FEATURE_FLAGS=charm-hub
$ juju bootstrap lxd test
$ juju deploy ubuntu
```

## Documentation changes

Architecture is now funneled through the cli and to the controller so we download
the correct charm binary blob.

### CODE REVIEW

File: api/charms/client.go [hunk 1]
```diff
@@ -95,6 +95,40 @@ func (c *Client) ResolveCharms(charms []CharmToResolve) ([]ResolvedCharm, error)
 	return resolvedCharms, nil
 }
 
+// DownloadInfo holds the information for a given DownloadInfo.
```
Comment (manadart) [inline api/charms/client.go:98]: Tautological?

File: charmhub/download.go [hunk 2]
```diff
@@ -96,6 +96,16 @@ func (c *DownloadClient) DownloadAndRead(ctx context.Context, resourceURL *url.U
...
 }
 
+// DownloadAndReadBundle returns a bundle archive retrieved from the given URL.
+func (c *DownloadClient) DownloadAndReadBundle(ctx context.Context, resourceURL *url.URL, archivePath string) (*charm.BundleArchive, error) {
+	err := c.Download(ctx, resourceURL, archivePath)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return charm.ReadBundleArchive(archivePath)
```
Comment (manadart) [inline charmhub/download.go:106]: Trace.

File: charmhub/refresh.go [hunk 3]
```diff
@@ -35,6 +31,24 @@ const (
 	RefreshAction Action = "refresh"
 )
 
+// RefreshPlatform defines a platform refreshing charms.
```
Comment (manadart) [inline charmhub/refresh.go:34]: "for refreshing charms"?

File: charmhub/refresh_test.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ func (s *RefreshSuite) TestRefresh(c *gc.C) {
 			Platform: transport.RefreshRequestPlatform{
 				OS:           "ubuntu",
 				Series:       "focal",
-				Architecture: "all",
+				Architecture: "amd64",
```
Comment (manadart) [inline charmhub/refresh_test.go:42]: We could use the default arch constant for all of these.

File: charmhub/selector/package_test.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package selector
```
Comment (manadart) [inline charmhub/selector/package_test.go:4]: Yes.

File: charmhub/selector/selector.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+		return nil, charm.Origin{}, errors.Trace(err)
+	}
+
+	channelMap = filterInfoChannelMap(channelMap, filterFn)
+	revisions, found := locateRevisionByChannel(c.sortInfoChannelMap(channelMap), *channel)
+	if !found {
+		if series != "" {
+			return nil, charm.Origin{}, errors.Errorf("%s %q not found for %q channel matching %q series", info.Type, info.Name, channel, series)
+		}
+		return nil, charm.Origin{}, errors.Errorf("%s %q not found with in the channel %q", info.Type, info.Name, channel)
```
Comment (manadart) [inline charmhub/selector/selector.go:102]: "within"

File: charmhub/selector/selector.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+type channelMapIndex struct {
+	witnessed  bool
+	channelMap transport.InfoChannelMap
+}
+
+var channelRisks = []string{"stable", "candidate", "beta", "edge"}
+
+func linkClosedChannels(channelMaps []transport.InfoChannelMap) ([]transport.InfoChannelMap, error) {
```
Comment (manadart) [inline charmhub/selector/selector.go:222]: I wonder if we should increase cross-team collaboration to see if there are now common snap/charmhub libs that might be extracted for use across code bases.

File: charmhub/selector/selector_test.go [hunk 8]
```diff
@@ -0,0 +1,961 @@
...
+
+	jujutesting "github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/charmhub/transport"
+	"github.com/juju/juju/core/charm"
+)
+
+type selectorSuite struct {
```
Comment (manadart) [inline charmhub/selector/selector_test.go:17]: Use `IsolationSuite` to get logger and sundries?

File: cmd/juju/application/deploy.go [hunk 9]
```diff
@@ -292,8 +316,13 @@ type DeployCommand struct {
 	// NewCharmRepo stores a function which returns a charm store client.
 	NewCharmRepo func() (*store.CharmStoreAdaptor, error)
 
+	// NewDownloadClient stores a function for getting a charm/bundle.
+	NewDownloadClient func() (store.DownloadBundleClient, error)
+
+	NewModelConfigClient func(base.APICallCloser) ModelConfigClient
```
Comment (manadart) [inline cmd/juju/application/deploy.go:322]: Might as well comment this one too.

File: cmd/juju/application/deployer/deployer_test.go [hunk 10]
```diff
@@ -397,7 +397,7 @@ func (s *deployerSuite) expectModelType() {
 }
 
 func (s *deployerSuite) expectGetBundle(err error) {
-	s.resolver.EXPECT().GetBundle(gomock.AssignableToTypeOf(&charm.URL{}), gomock.Any()).Return(s.bundle, err)
+	s.resolver.EXPECT().GetBundle(gomock.AssignableToTypeOf(&charm.URL{}), gomock.Any(), gomock.Any()).Return(s.bundle, err)
```
Comment (manadart) [inline cmd/juju/application/deployer/deployer_test.go:400]: Interesting. I've never used this matcher.

File: cmd/juju/application/store/store_test.go [hunk 11]
```diff
@@ -34,15 +34,14 @@ func (s *storeSuite) TestAddCharmFromURLAddCharmSuccess(c *gc.C) {
 
 	curl, err := charm.ParseURL("cs:testme")
 	c.Assert(err, jc.ErrorIsNil)
-	origin, err := utils.DeduceOrigin(curl, corecharm.Channel{Risk: corecharm.Beta})
+	origin, err := utils.DeduceOrigin(curl, corecharm.Channel{Risk: corecharm.Beta}, corecharm.Platform{Architecture: "amd64"})
```
Comment (manadart) [inline cmd/juju/application/store/store_test.go:37]: Const here too.

**Review by manadart — APPROVED** [review]
Some nitty suggestions, but great job. QA all good.

File: api/charms/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply api/charms/client.go:98]: I really struggle with basic docs sometimes :laughing:

File: charmhub/selector/selector.go (hunk 7 above)
Comment (SimonRichardson) [reply charmhub/selector/selector.go:222]: This is essentially only for bundles, which they don't require. They get to use the Refresh API and only that. We have to jump through hoops to get bundles to work in the same way as charms.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**